storage/remote/meta_shards.go

```diff
@@ -0,0 +1,274 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package remote
+
+import (
+	"context"
+	"strconv"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/cespare/xxhash"
+	"github.com/go-kit/kit/log/level"
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/snappy"
+	"github.com/prometheus/prometheus/pkg/textparse"
+	"github.com/prometheus/prometheus/prompb"
+	"github.com/prometheus/prometheus/scrape"
+)
+
+const (
+	metadataFlushDeadline = 1 * time.Minute // 2 times the metadata lookup
+	metadataMinBackoff    = 30 * time.Second
+	metadataMaxBackoff    = 5 * time.Minute
+)
+
+type metaShards struct {
```

Comment:
Why do we need to parallelise the sending of metadata?

Reply:
I had this already done by the time we decided not to do so - figured I would put it up for discussion as the non-parallelization would be much simpler to execute upon.

Reply:
Let do with non-parallelised to start.  We can add it later if we need it.